src/Microsoft.Maintainability.Analyzers/Core/AvoidUnusedPrivateFields.cs

```diff
@@ -32,7 +35,44 @@ public abstract class AvoidUnusedPrivateFieldsAnalyzer : DiagnosticAnalyzer
 
         public override void Initialize(AnalysisContext analysisContext)
         {
-            
+            analysisContext.RegisterCompilationStartAction(
+                (compilationContext) =>
+                {
+                    HashSet<IFieldSymbol> unreferencedPrivateFields = new HashSet<IFieldSymbol>();
+                    HashSet<IFieldSymbol> referencedPrivateFields = new HashSet<IFieldSymbol>();
+
+                    compilationContext.RegisterSymbolAction(
+                        (symbolContext) =>
+                        {
+                            IFieldSymbol field = (IFieldSymbol)symbolContext.Symbol;
+                            if (field.DeclaredAccessibility == Accessibility.Private && !referencedPrivateFields.Contains(field))
+                            {
+                                unreferencedPrivateFields.Add(field);
+                            }
+                        },
+                        SymbolKind.Field);
+
+                    compilationContext.RegisterOperationAction(
+                        (operationContext) =>
+                        {
+                            IFieldSymbol field = ((IFieldReferenceExpression)operationContext.Operation).Field;
+                            if (field.DeclaredAccessibility == Accessibility.Private)
+                            {
+                                referencedPrivateFields.Add(field);
+                                unreferencedPrivateFields.Remove(field);
+                            }
+                        },
+                        OperationKind.FieldReferenceExpression);
+
```

Comment:
It does with the current analyzer driver implementation but their is no guarantee provided by analyzer API semantics that guarantees it - basically we are free to change it in future.

The only ordering guarantees that are provided are:

1. If you have start action, then its nested actions are invoked after start action finishes.
2. An end action will always execute non-concurrently at the end of all non-end actions for the analysis unit (compilation, code block, operation block, etc.)